### PR TITLE
[core] Introduce stats in snapshot

### PR DESCRIPTION
**Pull request by Zouxxyy**
### Purpose

First step of [PIP-14: Paimon statistics](https://cwiki.apache.org/confluence/display/PAIMON/PIP-14%3A+Paimon+statistics)

snapshot-x
```json
{
  "version" : 3,
  "id" : 2,
  "schemaId" : 3,
  "baseManifestList" : "manifest-list-485c5847-cb20-4324-b2ea-fd73959c5857-0",
  "deltaManifestList" : "manifest-list-485c5847-cb20-4324-b2ea-fd73959c5857-1",
  "changelogManifestList" : null,
  "commitUser" : "e30257ac-a617-4571-a57b-60c3dd608754",
  "commitIdentifier" : 9223372036854775807,
  "commitKind" : "ANALYZE",
  "timeMillis" : 1704944283486,
  "logOffsets" : { },
  "totalRecordCount" : 6,
  "deltaRecordCount" : 0,
  "changelogRecordCount" : 0,
  "watermark" : null,
  "statistics" : "stats-87effd5d-48fd-4aab-81fe-4222b847d247-0"
}
```
statistics/stats-87effd5d-48fd-4aab-81fe-4222b847d247-0
```json
{
  "snapshotId": 2,
  "mergedRecordCount" : 10,
  "mergedRecordSize" : 1000,
  "colStats" : {
    "orderId" : {
      "distinctCount" : 10,
      "min" : "1",
      "max" : "10",
      "nullCount" : 0,
      "avgLen" : 8,
      "maxLen" : 8
    }
  }
}
```

### API and Format

write
```
FileStoreCommitImpl commit = store.newCommit();
commit.writeStats(stats, Long.MAX_VALUE);
```

read
```
StatsFileHandler statsFileHandler = store.newStatsFileHandler();
Optional<Stats> stats = statsFileHandler.readStats();
```

### Next

Future work

- expire of stats
- read stats with system table
- spark analyze

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/Snapshot.java [hunk 1]
```diff
@@ -184,7 +192,8 @@ public Snapshot(
             @Nullable Long totalRecordCount,
             @Nullable Long deltaRecordCount,
             @Nullable Long changelogRecordCount,
-            @Nullable Long watermark) {
+            @Nullable Long watermark,
+            @Nullable String statistics) {
```
Comment (YannByron) [inline paimon-core/src/main/java/org/apache/paimon/Snapshot.java:196]: If the partition-level statistics are supported, how to retrieve those?

File: paimon-core/src/main/java/org/apache/paimon/stats/ColStats.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+ * <ul>
+ *   <li>distinctCount: the number of distinct values
+ *   <li>min: the minimum value of the column
+ *   <li>max: the maximum value of the column
+ *   <li>nullCount: the number of nulls
+ *   <li>avgLen: average column length
+ *   <li>maxLen: max column length
+ * </ul>
+ */
+public class ColStats {
```
Comment (YannByron) [inline paimon-core/src/main/java/org/apache/paimon/stats/ColStats.java:48]: can we define this class with the dataType of column, like:
`
class ColStats[T]
`
T is one of the implementations of `org.apache.paimon.types.DataType`.

File: paimon-core/src/main/java/org/apache/paimon/stats/ColStats.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        return Objects.hash(
+                distinctCount, serializedMin, min, serializedMax, max, nullCount, avgLen, maxLen);
+    }
+
+    @Override
+    public String toString() {
+        return "ColStats{"
+                + "distinctCount="
+                + distinctCount
+                + ", serializedMin='"
```
Comment (YannByron) [inline paimon-core/src/main/java/org/apache/paimon/stats/ColStats.java:194]: `min` and `max` is ok.

File: paimon-core/src/main/java/org/apache/paimon/stats/Stats.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    private static final String FIELD_MERGED_RECORD_COUNT = "mergedRecordCount";
+    private static final String FIELD_MERGED_RECORD_SIZE = "mergedRecordSize";
+    private static final String FIELD_COL_STATS = "colStats";
+
+    @JsonProperty(FIELD_SNAPSHOT_ID)
+    private final long snapshotId;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    @JsonProperty(FIELD_MERGED_RECORD_COUNT)
+    private final @Nullable Long mergedRecordCount;
```
Comment (YannByron) [inline paimon-core/src/main/java/org/apache/paimon/stats/Stats.java:61]: If `Stats` is created, `mergedRecordCount` and `mergedRecordSize` must be non-nullable. Remove `@Nullable`.

File: paimon-core/src/main/java/org/apache/paimon/stats/Stats.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    @JsonProperty(FIELD_MERGED_RECORD_SIZE)
+    private final @Nullable Long mergedRecordSize;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    @JsonProperty(FIELD_COL_STATS)
+    private final @Nullable Map<String, ColStats> colStats;
```
Comment (YannByron) [inline paimon-core/src/main/java/org/apache/paimon/stats/Stats.java:69]: Can we init `colStats` an empty map if no col stats?

File: paimon-core/src/main/java/org/apache/paimon/stats/StatsFile.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+import org.apache.paimon.fs.FileIO;
+import org.apache.paimon.fs.Path;
+import org.apache.paimon.utils.PathFactory;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+/** Stats file contains stats. */
+public class StatsFile {
```
Comment (YannByron) [inline paimon-core/src/main/java/org/apache/paimon/stats/StatsFile.java:29]: According to OOP, these methods(`read`, `write`, `delete` and `exists`) should not be the members of `StatsFile`.

File: paimon-core/src/main/java/org/apache/paimon/Snapshot.java (hunk 1 above)
Comment (Zouxxyy) [reply paimon-core/src/main/java/org/apache/paimon/Snapshot.java:196]: Plan to add a new field named `partitionStatistics` (a manifestList link) when going to support partition-level statistics

File: paimon-core/src/main/java/org/apache/paimon/stats/StatsFile.java (hunk 6 above)
Comment (Zouxxyy) [reply paimon-core/src/main/java/org/apache/paimon/stats/StatsFile.java:29]: just follow the implement of `indexFile` and `indexFileHandle`

**Comment by TaoZex** [comment]
In pip-14, we included statistics for maxLen and avgLen, would you consider adding minLen？

**Comment by Zouxxyy** [comment]
> In pip-14, we included statistics for maxLen and avgLen, would you consider adding minLen？

The primary purpose of PIP-14 is to enable Paimon to preserve stats for query engine CBO (especially for Spark), Below are the stats collected by various engines, the common ones among them are the stats introduced in this PIP.

Hive: https://cwiki.apache.org/confluence/display/Hive/StatsDev
Flink: https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/analyze/
Spark: https://spark.apache.org/docs/latest/sql-ref-syntax-aux-analyze-table.html

Besides, from what I know, aside from histograms (maybe will add to paimon in the future), distinctCount, min, max, avglen, and nullCount are useful in cost estimation, maxLen is hardly used in Spark(open source)'s CBO. Nevertheless, we have decided to include it in order to maintain consistency with Spark. Of course, if we find other certain stats are very useful, we can also add them.

Cost estimation example: https://cwiki.apache.org/confluence/display/Hive/Cost-based+optimization+in+Hive.

**Comment by TaoZex** [comment]
> > In pip-14, we included statistics for maxLen and avgLen, would you consider adding minLen？
> 
> The primary purpose of PIP-14 is to enable Paimon to preserve stats for query engine CBO (especially for Spark), Below are the stats collected by various engines, the common ones among them are the stats introduced in this PIP.
> 
> Hive: https://cwiki.apache.org/confluence/display/Hive/StatsDev Flink: https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/analyze/ Spark: https://spark.apache.org/docs/latest/sql-ref-syntax-aux-analyze-table.html
> 
> Besides, from what I know, aside from histograms (maybe will add to paimon in the future), distinctCount, min, max, avglen, and nullCount are useful in cost estimation, maxLen is hardly used in Spark(open source)'s CBO. Nevertheless, we have decided to include it in order to maintain consistency with Spark. Of course, if we find other certain stats are very useful, we can also add them.
> 
> Cost estimation example: https://cwiki.apache.org/confluence/display/Hive/Cost-based+optimization+in+Hive.

Thanks for your reply, which helped me understand the design idea of this part.

File: paimon-core/src/main/java/org/apache/paimon/Snapshot.java (hunk 1 above)
Comment (YannByron) [reply paimon-core/src/main/java/org/apache/paimon/Snapshot.java:196]: both `statistics` and `partitionStatistics` are related to statistics, why we split into two?

File: paimon-core/src/main/java/org/apache/paimon/stats/Stats.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    @JsonProperty(FIELD_MERGED_RECORD_COUNT)
+    private final @Nullable Long mergedRecordCount;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    @JsonProperty(FIELD_MERGED_RECORD_SIZE)
+    private final @Nullable Long mergedRecordSize;
+
+    @JsonProperty(FIELD_COL_STATS)
+    private final Map<String, ColStats<?>> colStats;
```
Comment (YannByron) [inline paimon-core/src/main/java/org/apache/paimon/stats/Stats.java:69]: We need to take `schema evolution` into consideration. So the `key` of this map should contain the id of column. Therefore, the version of schema which is used in this statistics should be kept.

**Comment by Zouxxyy** [comment]
two update:
- add schemaId in stats: snapshot may expire, but schema never
- add colId for each col in colStats: for schema evolution in the future

**Comment by YannByron** [comment]
+1

**Review by JingsongLi — APPROVED** [review]
+1

**Comment by Zouxxyy** [comment]
#2799

**Comment by YannByron** [comment]
https://github.com/apache/incubator-paimon/issues/2404

**Pull request merged**